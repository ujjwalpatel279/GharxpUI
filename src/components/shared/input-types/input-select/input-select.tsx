import React, { ChangeEvent, forwardRef, useEffect } from 'react';
import { useForm } from 'react-hook-form';

import './input-select.scss';
interface OptionsInterface {
  id: string | number;
  value: string;
  displayText?: string;
  selectable?: boolean;
}

interface InputSelectProps {
  id: string;
  options: OptionsInterface[];
  forceSelection?: boolean;
  required: boolean;
  disabled?: boolean;
  error?: boolean;
  setValue?: ReturnType<typeof useForm>['setValue'];
  value?: string | number | boolean;
  onChange?: (e: ChangeEvent<HTMLSelectElement | HTMLInputElement>) => void;
}

type Ref = HTMLSelectElement;

export const InputSelect = forwardRef<Ref, InputSelectProps>(
  ({ id, options, forceSelection, required, disabled, error, onChange, setValue, value }, ref) => {
    let selectClasses = 'c-form-select';

    useEffect(() => {
      if (value && setValue) {
        setValue(id, value);
      }
    }, [value]);

    if (error) {
      selectClasses += ' isInvalid';
    }

    if (required) {
      selectClasses += ' isRequired';
    }

    return (
      <div className="c-input-select-wrapper">
        <select
          name={id}
          id={id}
          className={selectClasses}
          ref={ref}
          disabled={disabled}
          required={required}
          aria-invalid={error ? 'true' : 'false'}
          data-testid="input-select"
          onChange={onChange}
        >
          {forceSelection && <option value="">Please select</option>}
          {options.map((option: OptionsInterface) => (
            <option key={option.id} value={option.id} disabled={option.selectable ?? true ? false : true}>
              {option.displayText ?? option.value}
            </option>
          ))}
        </select>
      </div>
    );
  },
);

InputSelect.displayName = 'InputSelect';
