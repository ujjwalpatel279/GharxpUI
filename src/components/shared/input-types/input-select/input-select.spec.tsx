import React from 'react';
import { render, cleanup } from '@testing-library/react';

import { InputSelect } from './input-select';

const mockOptions = [
  {
    id: '1',
    value: 'One',
  },
  {
    id: '2',
    value: 'Two',
  },
];

describe('FormInput Component', () => {
  afterEach(cleanup);

  it('renders the form select component', () => {
    const { container } = render(
      <InputSelect id="testSelect" required={false} options={mockOptions} ref={() => null} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('aria-invalid is false', () => {
    const { getByTestId } = render(
      <InputSelect id="testSelect" options={mockOptions} required={false} error={false} ref={() => null} />,
    );

    expect(getByTestId('input-select').getAttribute('aria-invalid')).toBe('false');
  });

  it('aria-invalid is true', () => {
    const { getByTestId } = render(
      <InputSelect id="testSelect" options={mockOptions} required={false} error={true} ref={() => null} />,
    );

    expect(getByTestId('input-select').getAttribute('aria-invalid')).toBe('true');
  });

  it('first option is "Please select"', () => {
    const { getByTestId } = render(
      <InputSelect id="testSelect" forceSelection={true} required={false} options={mockOptions} ref={() => null} />,
    );

    const select = getByTestId('input-select') as HTMLInputElement;
    expect(select.value).toBe('');
  });

  it('first option is "One"', () => {
    const { getByTestId } = render(
      <InputSelect id="testSelect" required={false} options={mockOptions} ref={() => null} />,
    );

    const select = getByTestId('input-select') as HTMLInputElement;
    expect(select.value).toBe('1');
  });
});
