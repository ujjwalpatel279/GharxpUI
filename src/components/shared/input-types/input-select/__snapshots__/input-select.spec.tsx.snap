// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput Component renders the form select component 1`] = `
<div
  class="c-input-select-wrapper"
>
  <select
    aria-invalid="false"
    class="c-form-select"
    data-testid="input-select"
    id="testSelect"
    name="testSelect"
  >
    <option
      value="1"
    >
      One
    </option>
    <option
      value="2"
    >
      Two
    </option>
  </select>
</div>
`;
